### PR TITLE
Forbid writing to cluster-managed main on restart

### PR DESCRIPTION
**Pull request by as51340**
When managed by coordinator, main should't be writeable. Only after receiving RPC for accepting messages, main can start accepting write queries again. RPC will be sent if after main restart, main will stay main and not become replica.

[master < Epic] PR
- [ ] Check, and update documentation if necessary
- [ ] Write E2E tests
- [ ] Compare the [benchmarking results](https://bench-graph.memgraph.com/) between the master branch and the Epic branch
- [ ] Provide the full content or a guide for the final git message

[master < Task] PR
- [ ] Check, and update documentation if necessary
- [ ] Provide the full content or a guide for the final git message

To keep docs changelog up to date, one more thing to do:
- [ ] Write a release note here, including added/changed clauses
- [ ] Tag someone from docs team in the comments

### CODE REVIEW

**Review by andrejtonev — COMMENTED** [review]
Good work!
Just 2 notes:
1) We should set the enable_write to false on startup/restart
2) This is not thread safe, right?

**Comment by as51340** [comment]
By thread safe you mean, one thread can set it to true from rpc and other could just be initializing it or?

**Comment by andrejtonev** [comment]
> By thread safe you mean, one thread can set it to true from rpc and other could just be initializing it or?

We have multiple threads that could be reading/writing this variable. The RPC is setting it and all query execution threads are reading it. My comment was me basically just checking that replication state is still not thread safe and that we need to fix that.

**Comment by as51340** [comment]
> We have multiple threads that could be reading/writing this variable. The RPC is setting it and all query execution threads are reading it. My comment was me basically just checking that replication state is still not thread safe and that we need to fix that.


Yes, agree. I can put this under atomic but still whole state isn't thread safe. You want me to put writing_enabled under atomic before merging this?

**Comment by andrejtonev** [comment]
> Yes, agree. I can put this under atomic but still whole state isn't thread safe. You want me to put writing_enabled under atomic before merging this?

I think it's fine for now. We need to fix the whole state. Just set it to false on startup and I think that's it.

**Comment by as51340** [comment]
> I think it's fine for now. We need to fix the whole state. Just set it to false on startup and I think that's it.

It already is false on startup I think

File: src/coordination/coordinator_handlers.cpp [hunk 1]
```diff
@@ -148,7 +154,7 @@ void CoordinatorHandlers::PromoteReplicaToMainHandler(replication::ReplicationHa
       }
     }
   }
-  spdlog::error(fmt::format("FICO : Promote replica to main was success {}", std::string(req.main_uuid_)));
+  spdlog::info("Promote replica to main was success {}", std::string(req.main_uuid_));
```
Comment (antoniofilipovic) [inline src/coordination/coordinator_handlers.cpp:157]: Nice

File: src/utils/functional.hpp [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+// licenses/APL.txt.
+
+#pragma once
+
+#include <algorithm>
+#include <vector>
+
+namespace memgraph::utils {
+
+template <class F, class T, class R = typename std::result_of<F(T)>::type, class V = std::vector<R>>
```
Comment (antoniofilipovic) [inline src/utils/functional.hpp:19]: Can you use here std::result_of_t?

**Review by antoniofilipovic — COMMENTED** [review]
Only one comment, looks good

**Comment by andrejtonev** [comment]
> It already is false on startup I think

Ah in Fetch it's false. Okay!

**Review by andrejtonev — APPROVED** [review]

**Review by antoniofilipovic — APPROVED** [review]

File: src/utils/functional.hpp (hunk 2 above)
Comment (as51340) [reply src/utils/functional.hpp:19]: will change

**Comment by as51340** [comment]
@kgolubic Release note: To avoid possible split brain issue and data inconsistencies, writing on main is by default forbidden after it gets restarted until the coordinator explicitly allows old main to be writeable. This will only happen if there was no new main in the cluster in the time between main failed and came back.

**Pull request merged**